plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+            key = key.substring(PATH_SEPARATOR.length());
+        }
+        if (key.endsWith(PATH_SEPARATOR)) {
+            key = key.substring(0, key.length() - PATH_SEPARATOR.length());
+        }
+        return key;
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static AwsCredentialsProvider createAwsCredentialsProvider(ExchangeS3Config config)
+    {
+        if (config.getS3AwsAccessKey() != null && config.getS3AwsSecretKey() != null) {
+            return StaticCredentialsProvider.create(AwsBasicCredentials.create(config.getS3AwsAccessKey(), config.getS3AwsSecretKey()));
+        }
+        return DefaultCredentialsProvider.create();
+    }
+
+    private S3Client createS3Client(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3ClientBuilder clientBuilder = S3Client.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3AsyncClientBuilder clientBuilder = S3AsyncClient.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static class S3ExchangeStorageWriter
+            implements ExchangeStorageWriter
+    {
+        private final S3Client s3Client;
+        private final S3AsyncClient s3AsyncClient;
+        private final String bucketName;
+        private final String key;
+        private final int partSize;
+        private final Optional<SecretKey> secretKey;
+
+        private int currentPartNumber;
+        private Optional<String> uploadId = Optional.empty();
+        private final List<CompletableFuture<CompletedPart>> uploadFutures = new ArrayList<>();
+
+        public S3ExchangeStorageWriter(S3Client s3Client, S3AsyncClient s3AsyncClient, String bucketName, String key, int partSize, Optional<SecretKey> secretKey)
+        {
+            this.s3Client = requireNonNull(s3Client, "s3Client is null");
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public ListenableFuture<Void> write(Slice slice)
```

Comment:
Currently the interface allows initiating multiple writes to be executed in parallel. At first it seemed a little weird, but actually it looks like something that is good for what we are doing, as it allows to handle skewed writes pretty well. 

There are still two concerns from my side:

1. Is there a limit enforced by AWS on how many parts can be uploaded in parallel? Should we have a back pressure mechanism in place in case there is?
2. I wonder whether other distributed file systems allow parallel multipart upload? Maybe we don't have to bother about it now, since we are trying to focus on S3, but it is something just to keep in mind.

Reply:
1. From https://stackoverflow.com/questions/35184353/is-there-a-maximum-concurrency-for-aws-s3-multipart-uploads, the maximum concurrency is 10K, which doesn't seem to be a potential issue in practice. 10K is also the upper bound of the parts we can upload, this means if each part is of size 5MB, then each spooling file is at most 5GB. This may expose limits on use cases.
2. Yes, Microsoft and Google have similar things https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-block-blobs and https://cloud.google.com/storage/docs/multipart-uploads